### DISCUSSION TITLE
_get_column_info on PGDialect replacement with 2.0

### DISCUSSION THREAD

**Discussion by jasonamyers** [sqlalchemy/sqlalchemy#9259] [Usage Questions] [answered]
Prior to the release of SQLAlchemy 2.0, sqlalchemy-redshift took advantage of the _get_column_info via a super call. [Code Link](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/blob/b8d17be6dde4d860584c13ec6c4298a2c498d062/sqlalchemy_redshift/dialect.py#L799) What is the proper replacement for this going forward?

**Comment by CaselIT** [comment]
Hi,

v2 rewrote the reflection logic for postgresql to use a new multi table approach. I don't know how compatible redshift is with postgresql pg_catalog, so it may need to vendor some of the parts that were in 1.4 if the current logic does not work with it.
Anyway the direct replacement for `_get_column_info ` is https://github.com/sqlalchemy/sqlalchemy/blob/c40f529b0b93f832595821de96b96f3795a3a384/lib/sqlalchemy/dialects/postgresql/base.py#L3541 that works on a list of dicts spanning over multiple tables

edited to add: 3rd party dialect can opt into the new multi table approach, but sqlalchemy has fallbacks in place to use the 1 series methods. For example mysql or mssql reflection logic is pretty much identical between the v2 and v1.4

**Comment by zzzeek** [comment, marked as answer]
Hi Jason -

The reflection code for PostgreSQL is [totally new in 2.0](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#major-architectural-performance-and-api-enhancements-for-database-reflection) so if redshift has  their own reflection that nonetheless calls upon _get_column_info(), they should likely vendor the code from the 1.4.x series, or if they prefer they can migrate up to the newer reflection code at some point.   Re: vendoring _get_column_info(), it's MIT licensed so there is no restriction.

**Comment by jasonamyers** [comment]
@CaselIT  @zzzeek, thanks. I'll vendor it, for now, to get it supported and then try to work out the new 2.0 reflection methodology updates we'd need.

**Reply by CaselIT** [reply]
note that there are no plan on deprecating the old ones, so it's not required for 3rd party dialect to adopt them.
the new version were added since it was otherwise not possible to solve the performance issues that have been reported multiple times in the reflection system

**Reply by jasonamyers** [reply]
All of that is good and fair, but the 2.0 changes and the way sqlalchemy-redshift currently works makes sqlalchemy-redshift incompatible with SQ 2.0 right now, and if we need to put the effort in, I'd rather put the effort towards the latest usage patterns.

**Reply by CaselIT** [reply]
yes, if it used as base the postgresql dialect, that had a lot of changes around the reflection, that was basically rewritten in db08a699489c9b0259579d7ff7fd6bf3496ca3a2
